packages/arb-ts/src/lib/L1ToL2Message/L1ToL2Message.ts

```diff
@@ -0,0 +1,302 @@
+    this.l2TicketCreationTxnHash = l2TicketCreationTxnHash
+    this.arbRetryableActions = new ArbRetryableActions(signersAndProviders)
+    this.messageNumber = messageNumber
+    this.l1TxnHash = l1TxnHash
+  }
+  async initFromL1Txn(
+    signersAndProviders: SignersAndProviders,
+    l1Txn: string | TransactionReceipt,
+    messageNumberIndex?: number
+  ): Promise<L1ToL2Message> {
+    const l1TxnReceipt = await getTxnReceipt(
+      l1Txn,
+      signersAndProviders.l1Provider
+    )
+    const l1TxnHash = l1TxnReceipt.transactionHash
+
+    const messageNumbers = getMessageNumbers(l1TxnReceipt)
+    if (messageNumbers === undefined)
+      throw new Error(`No l1 to L2 message found for ${l1TxnHash}`)
+
+    if (
+      messageNumberIndex !== undefined &&
+      messageNumberIndex > messageNumbers.length
+    )
+      throw new Error(
+        `Provided message number out of range for ${l1TxnHash}; index was ${messageNumberIndex}, but only ${messageNumbers.length} messages`
+      )
+    if (messageNumberIndex === undefined && messageNumbers.length > 1)
+      throw new Error(
+        `${messageNumbers.length} L2 messages for ${l1TxnHash}; must provide messamessageNumberIndex (or use initAllFromL1Txn)`
+      )
+    const messageNumber = messageNumbers[messageNumberIndex || 0]
+    if (!signersAndProviders.l2Provider) throw new Error('need l2 prov')
+    const chainID = (
+      await signersAndProviders.l2Provider.getNetwork()
+    ).chainId.toString()
+    const ticketCreationHash = calculateRetryableTicketCreationHash(
+      messageNumber,
+      BigNumber.from(chainID)
+    )
+    return new L1ToL2Message(
+      signersAndProviders,
+      ticketCreationHash,
+      messageNumber,
+      l1TxnHash
+    )
+  }
+
+  public async initAllFromL1Txn(
+    signersAndProviders: SignersAndProviders,
+    l1Txn: string | TransactionReceipt
+  ): Promise<L1ToL2Message[]> {
+    const l1TxnReceipt = await getTxnReceipt(
+      l1Txn,
+      signersAndProviders.l1Provider
+    )
+    const l1TxnHash = l1TxnReceipt.transactionHash
+    this.l1TxnHash = l1TxnHash
+
+    if (!signersAndProviders.l2Provider)
+      throw new Error('Missing required L2 Provider')
+    const chainID = (
+      await signersAndProviders.l2Provider.getNetwork()
+    ).chainId.toString()
+
+    const messageNumbers = getMessageNumbers(l1TxnReceipt)
+    if (!messageNumbers)
+      throw new Error('No l1 to l2 messages found in L1 txn ' + l1TxnHash)
+
+    return messageNumbers.map((msgNumber: BigNumber) => {
+      return new L1ToL2Message(
+        signersAndProviders,
+        calculateRetryableTicketCreationHash(
+          msgNumber,
+          BigNumber.from(chainID)
+        ),
+        msgNumber,
+        l1TxnHash
+      )
+    })
+  }
+
+  initFromL2Txn(
+    signersAndProviders: SignersAndProviders,
+    l2TicketCreationHash: string
+  ): L1ToL2Message {
+    return new L1ToL2Message(signersAndProviders, l2TicketCreationHash)
```

Comment:
hmm, we can't compute it directly, since the creation func is one-way, but the only variable is a message number, so we could just... save a big list? or "hash n' pray"? idk (it's not available in the L2 event logs unfortunately)